ospf6d/ospf6_bfd.c

```diff
@@ -44,72 +44,41 @@
 
 extern struct zclient *zclient;
 
-/*
- * ospf6_bfd_info_free - Free BFD info structure
- */
-void ospf6_bfd_info_free(void **bfd_info)
-{
-	bfd_info_free((struct bfd_info **)bfd_info);
-}
-
-/*
- * ospf6_bfd_show_info - Show BFD info structure
- */
-void ospf6_bfd_show_info(struct vty *vty, void *bfd_info, int param_only,
-			 json_object *json_obj, bool use_json)
-{
-	if (param_only)
-		bfd_show_param(vty, bfd_info, 1, 0, use_json, json_obj);
-	else
-		bfd_show_info(vty, bfd_info, 0, 1, use_json, json_obj);
-}
-
-/*
- * ospf6_bfd_reg_dereg_nbr - Register/Deregister a neighbor with BFD through
- *                           zebra for starting/stopping the monitoring of
- *                           the neighbor rechahability.
- */
-void ospf6_bfd_reg_dereg_nbr(struct ospf6_neighbor *on, int command)
-{
-	struct ospf6_interface *oi = on->ospf6_if;
-	struct interface *ifp = oi->interface;
-	struct bfd_info *bfd_info;
-	char src[64];
-	int cbit;
-
-	if (!oi->bfd_info || !on->bfd_info)
-		return;
-	bfd_info = (struct bfd_info *)oi->bfd_info;
-
-	if (IS_OSPF6_DEBUG_ZEBRA(SEND)) {
-		inet_ntop(AF_INET6, &on->linklocal_addr, src, sizeof(src));
-		zlog_debug("%s nbr (%s) with BFD",
-			   bfd_get_command_dbg_str(command), src);
-	}
-
-	cbit = CHECK_FLAG(bfd_info->flags, BFD_FLAG_BFD_CBIT_ON);
-
-	bfd_peer_sendmsg(zclient, bfd_info, AF_INET6, &on->linklocal_addr,
-			 on->ospf6_if->linklocal_addr, ifp->name, 0, 0, cbit,
-			 command, 0, ifp->vrf_id);
-
-	if (command == ZEBRA_BFD_DEST_DEREGISTER)
-		bfd_info_free((struct bfd_info **)&on->bfd_info);
-}
-
 /*
  * ospf6_bfd_trigger_event - Neighbor is registered/deregistered with BFD when
  *                           neighbor state is changed to/from 2way.
  */
 void ospf6_bfd_trigger_event(struct ospf6_neighbor *on, int old_state,
 			     int state)
 {
-	if ((old_state < OSPF6_NEIGHBOR_TWOWAY)
-	    && (state >= OSPF6_NEIGHBOR_TWOWAY))
-		ospf6_bfd_reg_dereg_nbr(on, ZEBRA_BFD_DEST_REGISTER);
-	else if ((old_state >= OSPF6_NEIGHBOR_TWOWAY)
-		 && (state < OSPF6_NEIGHBOR_TWOWAY))
-		ospf6_bfd_reg_dereg_nbr(on, ZEBRA_BFD_DEST_DEREGISTER);
+	int family;
+	struct in6_addr src, dst;
+
+	/* Skip sessions without BFD. */
+	if (on->bfd_session == NULL)
+		return;
+
+	if (old_state < OSPF6_NEIGHBOR_TWOWAY
+	    && state >= OSPF6_NEIGHBOR_TWOWAY) {
+		/*
+		 * Check if neighbor address changed.
+		 *
+		 * When the neighbor is configured BFD before having an existing
+		 * connection, then the destination address will be set to `::`
+		 * which will cause session installation failure. This piece of
+		 * code updates the address in that case.
+		 */
+		bfd_sess_addresses(on->bfd_session, &family, &src, &dst);
+		if (memcmp(&on->linklocal_addr, &dst, sizeof(dst))) {
+			bfd_sess_set_ipv6_addrs(on->bfd_session, &src,
+						&on->linklocal_addr);
+		}
+
+		bfd_sess_enable(on->bfd_session, true);
```

Comment:
Why do we need to change the "enable" flag here? This is the configurational value.

Reply:
The difference from PIM to OSPFv3 is that OSPFv3 has BFD information in the peer `show` commands.

In PIM if we have the BFD session data structure or not it does not matter because it is never shown (we can free the memory), however on OSPFv3 it displays the configuration and uptime.

The difference between `bfd_sess_uninstall` and `bfd_sess_enable(..., false)` is that the former will keep the session "disabled" (will be skipped/ignored) even on the case of daemon reconnection. If the session is only "uninstalled" it might get reinstalled if `zebra` or `bfdd` restarts. `bfd_sess_uninstall` is more of a action than a state and `bfd_sess_enable(..., false)` is the same as `uninstall` plus keeping it disabled.

```c
void bfd_sess_enable(struct bfd_session_params *bsp, bool enable)
{
	/* Remove the session when disabling. */
	if (!enable)
		_bfd_sess_remove(bsp);

	bsp->enabled = enable;
}
```

And here is why just uninstalling (`bfd_sess_uninstall`) is insufficient: https://github.com/FRRouting/frr/blob/master/lib/bfd.c#L1064

In conclusion: we disable the session so it gets uninstalled (and never gets installed again) and to keep the session data structure so we can keep track of current configuration + uptime.

I hope this answers this and the other questions.

Reply:
I am actually comparing this with OSPF implementation, not PIM, and the implementation is very different.
In `ospf_bfd_trigger_event` we only do install/uninstall without touching the enable flag.
If I understand your explanation correctly then we might have a problem in ospfd.

Reply:
I tested it locally and both `ospfd` and `ospf6d` delete the neighbor data structure along with `bfd_session` when session goes down... I did the extra effort in `ospf6d` but the gain was not that much.

When the session goes down the neighbor gets deleted and the session information doesn't get shown anyway. The only difference is: meanwhile the session doesn't go to state `Full` it shows the peer as `down`:

```
 Neighbor 10.254.254.5%r4-eth1
    Area 0.0.0.0 via interface r4-eth1 (ifindex 3)
    His IfIndex: 2 Link-local address: fe80::a44c:53ff:fe5e:daaf
    State Full for a duration of 00:00:00
    His choice of DR/BDR 10.254.254.5/0.0.0.0, Priority 1
    DbDesc status: Slave SeqNum: 0x254d0400
    Summary-List: 0 LSAs
    Request-List: 0 LSAs
    Retrans-List: 1 LSAs
      [Router Id:0.0.0.0 Adv:10.254.254.4]
    0 Pending LSAs for DbDesc in Time 00:00:00 [thread off]
    0 Pending LSAs for LSReq in Time 00:00:04 [thread on]
    0 Pending LSAs for LSUpdate in Time 00:00:04 [thread on]
    0 Pending LSAs for LSAck in Time 00:00:00 [thread off]
  BFD: Type: single hop
  Detect Multiplier: 3, Min Rx interval: 300, Min Tx interval: 300
  Status: Down, Last update: 0:00:00:01
```

Reply:
Fixed: BFD session data structure is always around except when BFD is not configured. Since `ospfd` and `ospf6d` does not keep neighbor data structures around when session goes away it doesn't make sense to do the extra state effort.